examples/jwt/src/main.rs

```diff
@@ -0,0 +1,178 @@
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::{convert::Infallible, fmt::Display, net::SocketAddr};
+
+// Quick instructions
+//
+// - get an authorization token:
+//
+// curl -s \
+//     -w '\n' \
+//     -H 'Content-Type: application/json' \
+//     -d '{"client_id":"foo","client_secret":"bar"}' \
+//     http://localhost:3000/auth
+//
+// - visit the protected area using the token for authentication
+//
+// curl -s \
+//     -w '\n' \
+//     -H 'Content-Type: application/json' \
+//     -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJiQGIuY29tIiwiY29tcGFueSI6IkFDTUUiLCJleHAiOjEwMDAwMDAwMDAwfQ.M3LAZmrzUkXDC1q5mSzFAs_kJrwuKz3jOoDmjJ0G4gM' \
+//     http://localhost:3000/protected
+//
+// - try to visit the protected area using an invalid token
+//
+// curl -s \
+//     -w '\n' \
+//     -H 'Content-Type: application/json' \
+//     -H 'Authorization: Bearer blahblahblah' \
+//     http://localhost:3000/protected
+
+static JWT_SECRET: &[u8] = b"secret";
+
+#[tokio::main]
+async fn main() {
+    // Set the RUST_LOG, if it hasn't been explicitly defined
+    if std::env::var("RUST_LOG").is_err() {
+        std::env::set_var("RUST_LOG", "example_jwt=debug")
+    }
+    tracing_subscriber::fmt::init();
+
+    let app = Router::new()
+        .route("/protected", get(protected))
+        .route("/auth", post(auth));
+
+    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+    tracing::debug!("listening on {}", addr);
+
+    axum::Server::bind(&addr)
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+}
+
+async fn protected(
+    TypedHeader(authorization): TypedHeader<Authorization<Bearer>>,
+) -> Result<String, AuthError> {
+    // Extract the authorization header value (i.e the token)
+    let token = authorization.0.token();
+
+    // Authenticate the user and returns his data
+    let validation = Validation {
+        sub: Some("b@b.com".to_string()),
+        ..Validation::default()
+    };
+    let token_data = decode::<Claims>(token, &DecodingKey::from_secret(JWT_SECRET), &validation)
+        .map_err(|_| AuthError::InvalidToken)?;
+
+    // Send the protected data to the user
+    Ok(format!(
+        "Welcome to the protected area :)\nYour data:\n{}",
+        token_data.claims
+    ))
+}
+
+async fn auth(Json(payload): Json<AuthPayload>) -> Result<Json<AuthBody>, AuthError> {
+    // Check if the user sent the credentials
+    if payload.client_id.is_empty() || payload.client_secret.is_empty() {
+        return Err(AuthError::MissingCredentials);
+    }
+    // Here you can check the user credentials from a database
+    if payload.client_id != "foo" || payload.client_secret != "bar" {
+        return Err(AuthError::WrongCredentials);
+    }
+
+    // Create the authorization token
+    let claims = Claims {
+        sub: "b@b.com".to_owned(),
+        company: "ACME".to_owned(),
```

Comment:
My JWT knowledge is a bit rusty. Are these fields data about the token issuer, or the user?

Reply:
Usually `sub` field contains some identification information like user id or uuid. So both fields are about user

Reply:
@Z4RX is right. The values `b@b.com`/`ACME` were extracted from the [`jsonwebtoken`'s example](https://github.com/Keats/jsonwebtoken/blob/master/examples/validation.rs#L15).